### ISSUE TITLE
-fno-omit-frame-pointer and save_return_addr

### ISSUE THREAD

**Issue by joabreu** [foss-for-synopsys-dwc-arc-processors/gcc#63]
Hi,

I was trying to use stack unwinding using -fno-omit-frame-pointer option and I verified that, for some functions, the blink register is not saved into stack which makes it not possible to find the return address. I verified that arc.c file has a variable called save_return_addr which enables/disables the saving of the blink register. I believe this register should always be saved when we are using -fno-omit-frame-pointer option, can you confirm this? Adding an "|| !flag_omit_frame_pointer" to the assigning of save_return_addr variable makes all functions save the return address, which is what is needed to correctly do stack unwinding.

Best regards,
Jose Miguel Abreu

**Comment by cupertinomiranda** [comment]
Hi Jose,

Our architecture does not always puts blink in the stack.
BLINK is only saved if any of the instructions in between the function epilogue and prologue, clobber it (i.e. changes it), or if the blink never needs to be restored (functions that never return).

Regarding unwinding we opted for a more optimal solution to create unwinding-tables in DWARF information (section .eh_frame). 
As not all applications require to (themselves) perform unwinding, this is not enabled by default and requires the use of -fasynchronous-unwind-tables option.

Nevertheless, unwinding in the current release has issues and will not always work if you do not use -fomit-frame-pointer.

I have recently worked in fixing the stack unwinder. I am sorry to say but doing what you suggested (as a fix) would be a cruelty to the tools.
The fix is much more complex. ;-)
Please expect a fix to the stack unwinding in the next release.

Hope it answers your questions.

Best regards (Atentamente),
Cupertino

**Comment by joabreu** [comment]
Hi Cupertino,

Thanks for your prompt answer!

I did it as a quick fix, just to see if it worked. This is not really blocking and its just for an hobby project. Please notice that I am not running Linux, I am running my own OS and I wanted to implement stacktrace without implementing a DWARF unwinding algorithm. As for now I don't have user space (i.e. all code runs in kernel mode). I tested the stacktrace code by calling the stacktrace function in the middle of the execution and after a "trap_s" exception, the unwinding worked correctly. I also looked at the assembly code and it seems fine, but I am quite new to ARC devel so I probably overlooked something. And I think stack corruption is occurring :(

Just to understand better, can you please expand in what can be the problems with this workaround?

And while at it: this is a simple function with the workaround:
arch_local_irq_enable c0f1           push_s       %blink
.1+0x02     1cfcb6c8       push         %fp
.1+0x06     230a3700       mov          %fp,%sp
.1+0x0a     c1a1           sub_s        %sp,%sp,4
.1+0x0c     226a0280       lr           %r2,[%status32]
.1+0x10     22450182       or           %r2,%r2,6
.1+0x14     20290080       flag         %r2
.1+0x18     1bfcb080       st           %r2,[%fp,-4]
.1+0x1c     78e0           nop_s        
.1+0x1e     240a36c0       mov          %sp,%fp
.1+0x22     1404341b       pop          %fp
.1+0x26     c0d1           pop_s        %blink
.1+0x28     7ee0           j_s          [%blink]
.1+0x2a     78e0           nop_s        

I don't really see the point in instruction .1+0x0a: Is this reserving space for the variables pushed before (it doesn't makes sense because I pushed two 32 bits vars, so it should be 8) or is this getting the call stack (this var has no input args)? And why does it in the remaining code we use %fp instead of %sp?

With best regards,
Jose Miguel Abreu

**Comment by cupertinomiranda** [comment]
Hi again,

Just to for the sake of resolution after our phone call.

The only problem with the workaround is that it starts to generate un-optimized code.
The push instruction both allocates space and stores the register value in the stack.
The instruction at 0x0a, allocates space in the stack for the upcoming store in 0x18.

Considering code correctness, this function could simply be :

lr %r2,[%status32]
or %r2,%r2,6
flag %r2
j_s [%blink]

It is generally preferable for size and speed to include some more complex unwinding mechanism (such as the one in libgcc), with the DWARF unwind-tables than un-optimizing the code by adding extra loads and stores to stack.

Cheers,
Cupertino

**Comment by vineetgarc** [comment]
My 2 cents - frame pointer usage is generally an "unoptimization" and frowned upon more so for the sake of stack unwinding. In ARC architecture, by default we aggressively omit frame pointer and instead rely on tooling such as dwarf.

Now for various reasons/constraints user could still  want -fno-omit-frame-pointer to work - which is fair ! But the point is even if saving blink works for simple cases, it breaks when callee registers (r13-r25) are also involved and generated by compiler for a certain function - the reason is ABI. 

So, in a Frame Pointer based ISA, the activation record needs to be 
1. push return address (BLINK on ARC) on stack (SP increments)
2. push caller's stack base (FP on ARC) on stack (SP increments)
3. set this frame's base (FP) to current SP
4. push Callee regs on stack (r13-r25 on ARC) - actual regs depends on function code.
5. create space for any more local vars (increments SP)

However ARC ABI (historically) mandates function prolog to have steps 1, 4, 2, 3. With that FP no longer contains the base address of current call frame, thus we can't use it as a deterministic anchor to get BLINK/FP of previous frame. In absence of dwarf info, debuggers such as gdb and other tools work by disassembling code. This is obviously non trivial to do.

Long story short, FP can't be used for unwinding and saving BLINK alone is not sufficient either.
